GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,216 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.Tracing
+{
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private const int DEFAULT_MAX_QUEUE_SIZE = 256;
+
+        private readonly Func<NamedPipeClientStream> createPipeFunc;
+        private readonly BlockingCollection<string> queue;
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+
+        // An integer is used here rather than the enum so that it can be used with Interlocked.Exchange(..)
+        private int state = (int)QueuedPipeStringWriterState.Unknown;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> createPipeFunc, int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE)
+        {
+            this.createPipeFunc = createPipeFunc;
+            this.queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public event EventHandler<QueuedPipeStringWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeStringWriterState
+        {
+            Unknown = 0,
+            Stopped = 1,
+            Failing = 2,
+            Healthy = 3,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
```

Comment:
What is the driver is checking for this.isDisposed in every call?

Reply:
It's good measure. You should not interact with an `IDisposable` instance once it's been disposed. Especially something like `Start`/`Stop`.

Reply:
We could call dispose right after this 'if' check right?  I guess how much protection does this give us?  Are there cases we could hit it?

Reply:
> We could call dispose right after this 'if' check right? 

You are correct. I'm saying (similar to the argument made in `JsonTracer`) that we're going to assume disposal and initalisation are single-threaded and 'correct', meaning we're going to gloss over it 😄 .. we only care about the TryEnqueue part.